src/main/java/com/vackosar/gitflowincrementalbuild/boundary/UnchangedProjectsRemover.java

```diff
@@ -258,10 +278,45 @@ private void printDelimiter() {
     }
 
     private Stream<MavenProject> streamProjectWithDownstreamProjects(MavenProject project, Configuration config) {
-        return Stream.concat(
-            Stream.of(project),
-            config.mavenSession.getProjectDependencyGraph().getDownstreamProjects(project, true).stream()
-                    .filter(proj -> !config.excludeDownstreamModulesPackagedAs.contains(proj.getPackaging())));
+        // note: not using computeIfAbsent() here since it would break recursive invocations with ConcurrentModificationException
+        Set<MavenProject> downstream = downstreamCache.get(project);
+        if (downstream == null) {
+            downstream = Stream
+                    .concat(Stream.of(project),
+                            config.mavenSession.getProjectDependencyGraph().getDownstreamProjects(project, true)
+                                    .stream().filter(proj -> isDownstreamModuleNotExcluded(proj, config)))
+                    .collect(Collectors.toCollection(LinkedHashSet::new));
+            if (PCKG_POM.equals(project.getPackaging())) {    // performance hint: bomArtifactIdRegex or similar could speed things up
+                downstream.addAll(findBOMDownstreamProjects(project, downstream, config));
+            }
+            downstreamCache.put(project, downstream);
+        }
+        return downstream.stream();
+    }
+
+    private Set<MavenProject> findBOMDownstreamProjects(MavenProject potentialBOMProject, Set<MavenProject> downstream, Configuration config) {
+        return config.mavenSession.getProjects().stream()
+                .filter(proj -> !downstream.contains(proj)) // optimization
+                .filter(proj -> isDownstreamModuleNotExcluded(proj, config))
+                .filter(proj -> Optional.ofNullable(proj.getOriginalModel())    // >original< model is crucial since BOM deps are gone in effective model
+                        .map(Model::getDependencyManagement)
+                        .map(depMgtm -> depMgtm.getDependencies().stream().anyMatch(dep -> isBOMImport(dep, potentialBOMProject, proj, config)))
+                        .orElse(false))
+                .flatMap(proj -> streamProjectWithDownstreamProjects(proj, config)) // (indirect) recursion!
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+    }
+
+    private boolean isDownstreamModuleNotExcluded(MavenProject proj, Configuration config) {
+        return !config.excludeDownstreamModulesPackagedAs.contains(proj.getPackaging());
+    }
+
+    private boolean isBOMImport(Dependency dependency, MavenProject potentialBOMProject, MavenProject depDefiningProject, Configuration config) {
+        LazyExpressionEvaluator evaluator = new LazyExpressionEvaluator(config.mavenSession, depDefiningProject);
+        return Objects.equals(evaluator.evaluate(dependency.getType()), PCKG_POM)
+                && Objects.equals(evaluator.evaluate(dependency.getScope()), "import")
```

Comment:
`import"` could also be extracted as a constant, just like `PCKG_POM`

Reply:
You can use `org.apache.maven.artifact.Artifact#SCOPE_IMPORT` if that is in the classpath